armbian/build
# rockchip-rk3588: Enable CONFIG_NETKIT kernel option

# Description

`netkit` is the Linux BPF-programmable network device. Netkit is used by [Cilium](https://cilium.io/) for High Performance Cloud Native Networking (CNI).
From the Cilium [docs](https://docs.cilium.io/en/stable/overview/intro/) what Cilium is.

> Cilium is open source software for transparently securing the network connectivity between application services deployed using Linux container management platforms like Docker and Kubernetes.

>At the foundation of Cilium is a new Linux kernel technology called eBPF, which enables the dynamic insertion of powerful security visibility and control logic within Linux itself. Because eBPF runs inside the Linux kernel, Cilium security policies can be applied and updated without any changes to the application code or container configuration.

Some links for a detailed description.

- https://lwn.net/Articles/949960/
- https://isovalent.com/blog/post/cilium-1-16/#h-cilium-netkit
- https://isovalent.com/blog/post/cilium-netkit-a-new-container-networking-paradigm-for-the-ai-era/#h-shortcut-4-introducing-netkit


# How Has This Been Tested?

- [x] ./compile.sh kernel BOARD=nanopc-cm3588-nas BRANCH=edge RELEASE=bookworm KERNEL_CONFIGURE=no BUILD_MINIMAL=yes BUILD_DESKTOP=no COMPRESS_OUTPUTIMAGE=img
- [x] Tested with nanopc-cm3588-nas, `modinfo netkit` shows built-in module info
- [x] Tested with a single Kubernetes instance that uses Cilium for networking and active traffic.

## Detailed test steps

- Check kernel module

        modinfo netkit
        name:           netkit
        filename:       (builtin)
        alias:          rtnl-link-netkit
        license:        GPL
        file:           drivers/net/netkit
        author:         Nikolay Aleksandrov <razor@blackwall.org>
        author:         Daniel Borkmann <daniel@iogearbox.net>
        description:    BPF-programmable network device

- spinup a Kubernetes instance (K3s)

        curl -sLS https://get.k3sup.dev | sh
        k3sup install --local --cluster --k3s-extra-args "--disable servicelb --disable traefik --flannel-backend=none --disable-network-policy --disable-kube-proxy --cluster-cidr=10.42.0.0/16,fd0:cafe:42::/56 --service-cidr=10.43.0.0/16,fd0:cafe:43::/112" --k3s-version v1.31.1+k3s1

- Install Cilium tools

        # fish shell
        set CILIUM_CLI_VERSION (curl -s https://raw.githubusercontent.com/cilium/cilium-cli/main/stable.txt)
        set CLI_ARCH arm64
        
        curl -L --fail --remote-name-all https://github.com/cilium/cilium-cli/releases/download/$CILIUM_CLI_VERSION/cilium-linux-$CLI_ARCH.tar.gz{,.sha256sum}
        shasum -a 256 -c cilium-linux-$CLI_ARCH.tar.gz.sha256sum
        tar xzvfC cilium-linux-$CLI_ARCH.tar.gz /usr/local/bin
        rm cilium-linux-$CLI_ARCH.tar.gz{,.sha256sum}

- Install Cilium

        set -x KUBECONFIG /etc/rancher/k3s/k3s.yaml
        set API_SERVER_IP 192.168.2.10
        set API_SERVER_PORT 6443
        
        cilium install --version 1.16.2 \
          --set k8sServiceHost=$API_SERVER_IP \
          --set k8sServicePort=$API_SERVER_PORT \
          --set routingMode=native \
          --set bpf.datapathMode=netkit \
          --set bpf.masquerade=true \
          --set bpf.ipv6-masquerade=true \
          --set ipv4.enabled=true \
          --set ipv6.enabled=true \
          --set enableIPv4BIGTCP=true \
          --set enableIPv6BIGTCP=true \
          --set autoDirectNodeRoutes=true \
          --set ipam.mode=kubernetes \
          --set ipv4NativeRoutingCIDR="10.0.0.0/8" \
          --set ipv6NativeRoutingCIDR="fd00::/48" \
          --set ipam.operator.clusterPoolIPv4MaskSize="24" \
          --set ipam.operator.clusterPoolIPv6MaskSize="112" \
          --set ipam.operator.clusterPoolIPv4PodCIDRList="10.42.0.0/16" \
          --set ipam.operator.clusterPoolIPv6PodCIDRList="fd0:cafe:42::/56" \
          --set kubeProxyReplacement=true \
          --helm-set=operator.replicas=1

- Check Cilium and K8s

Afer the status returns no errors, you should see all pods are running.

        root@uhu ~# cilium status --wait
            /¯¯\
         /¯¯\__/¯¯\    Cilium:             OK
         \__/¯¯\__/    Operator:           OK
         /¯¯\__/¯¯\    Envoy DaemonSet:    OK
         \__/¯¯\__/    Hubble Relay:       disabled
            \__/       ClusterMesh:        disabled

        DaemonSet              cilium             Desired: 1, Ready: 1/1, Available: 1/1
        DaemonSet              cilium-envoy       Desired: 1, Ready: 1/1, Available: 1/1
        Deployment             cilium-operator    Desired: 1, Ready: 1/1, Available: 1/1
        Containers:            cilium             Running: 1
                               cilium-envoy       Running: 1
                               cilium-operator    Running: 1
        Cluster Pods:          3/3 managed by Cilium
        Helm chart version:    1.16.2
        Image versions         cilium             quay.io/cilium/cilium:v1.16.2@sha256:4386a8580d8d86934908eea022b0523f812e6a542f30a86a47edd8bed90d51ea: 1
                               cilium-envoy       quay.io/cilium/cilium-envoy:v1.29.9-1726784081-a90146d13b4cd7d168d573396ccf2b3db5a3b047@sha256:9762041c3760de226a8b00cc12f27dacc28b7691ea926748f9b5c18862db503f: 1
                               cilium-operator    quay.io/cilium/operator-generic:v1.16.2@sha256:cccfd3b886d52cb132c06acca8ca559f0fce91a6bd99016219b1a81fdbc4813a: 1
        
        
        root@uhu ~# kubectl get pod -A
        NAMESPACE     NAME                                      READY   STATUS    RESTARTS   AGE
        kube-system   cilium-4474q                              1/1     Running   0          2m8s
        kube-system   cilium-envoy-qh44h                        1/1     Running   0          2m8s
        kube-system   cilium-operator-b5bd59ff7-2f9nv           1/1     Running   0          2m8s
        kube-system   coredns-56f6fc8fd7-pjzgs                  1/1     Running   0          2m31s
        kube-system   local-path-provisioner-846b9dcb6c-dtdz4   1/1     Running   0          2m31s
        kube-system   metrics-server-5985cbc9d7-7xhbd           1/1     Running   0          2m31s

- Create network traffic

        cilium connectivity test

Running this test or deploying anything else does not create any kernel panic, which means all is working well with the kernel module.

# Checklist:
